src/rewrites.rs

```diff
@@ -849,27 +1037,519 @@ impl MultiPatterns {
                 if compatible(&subst_1_dec, &subst_2_dec, &map_1.var_map) {
                     // If so, merge two substitutions
                     let merged_subst = merge_subst(subst_1_dec, subst_2_dec, &map_1.var_map);
+                    // Check if any source pattern contains blacklisted nodes
+                    if self.filter_after {
+                        if contains_blacklist(
+                            rule.0.ast.as_ref(),
+                            &mut runner.egraph,
+                            &merged_subst,
+                        )
+                        .0 || contains_blacklist(
+                            rule.1.ast.as_ref(),
+                            &mut runner.egraph,
+                            &merged_subst,
+                        )
+                        .0
+                        {
+                            continue;
+                        }
+                    }
 
                     // check_pat on both dst patterns
-                    if check_pat(rule.2.ast.as_ref(), &mut runner.egraph, &merged_subst).0 {
-                        if check_pat(rule.3.ast.as_ref(), &mut runner.egraph, &merged_subst).0 {
+                    let (valid_1, _, _, existing_1) = check_pat(
+                        rule.2.ast.as_ref(),
+                        &mut runner.egraph,
+                        &merged_subst,
+                        /*get_exist_nodes=*/ self.filter_after,
+                    );
+                    let (valid_2, _, _, existing_2) = check_pat(
+                        rule.3.ast.as_ref(),
+                        &mut runner.egraph,
+                        &merged_subst,
+                        /*get_exist_nodes=*/ self.filter_after,
+                    );
+                    if valid_1 && valid_2 {
+                        let cycle_check_passed = if self.no_cycle {
+                            if self.filter_after {
```

Comment:
Why is the name filter_after? I though you check for cycles before applying a rule.

Reply:
Yes I do pre-check before applying a rule. This filter_after refers to we have a post-processing step to add nodes to filtered list